### PR TITLE
Jest tests for utils

### PR DESCRIPTION
**Pull request by m-allanson**
> Which issue, if any, is this issue related to?

Refs PR #2083

> Is there anything in the PR that needs further explanation?

This is a PR to merge into the `new-test-runner` branch, not `master`.

### CODE REVIEW

File: src/utils/__tests__/atRuleParamIndex.test.js [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-    t.equal(atRuleParamIndex(atRule), 8)
+    expect(atRuleParamIndex(atRule)).toBe(8)
   })
   atRules("@import\n'x.css');", atRule => {
-    t.equal(atRuleParamIndex(atRule), 8)
+    expect(atRuleParamIndex(atRule)).toBe(8)
   })
   atRules("@document url-prefix(http://www.w3.org/Style/)", atRule => {
-    t.equal(atRuleParamIndex(atRule), 10)
+    expect(atRuleParamIndex(atRule)).toBe(10)
```
Comment (davidtheclark) [inline src/utils/__tests__/atRuleParamIndex.test.js]: We are only lucky if this works ... Each call to `atRules` is asynchronous, cannot be treated synchronously, like this.

Instead, I'd suggest something like this: 

- Modify the `atRules` util so it's less fancy. It should just return the Promise that resolves with a PostCSS result.
- Change the test to `describe` and put each assertion within its own `it` statement.
- For each assertion, return the Promise from `atRules`, following by `.then(result => /* do assertion here */)`.

File: src/utils/__tests__/beforeBlockString.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  expect(postcssCheck("a {}")).toBe("a ")
+  expect(postcssCheck("\na\n{}")).toBe("\na\n")
+  expect(postcssCheck("\n\na,\nb,\n\tspan > .foo\n{}")).toBe("\n\na,\nb,\n\tspan > .foo\n")
+})
+
+it("beforeBlockString at-rules", () => {
+  expect(postcssCheck("@media print {}")).toBe("@media print ")
+  expect(postcssCheck("\n@media print, screen\n\t{}")).toBe("\n@media print, screen\n\t")
+  expect(postcssCheck("@supports (animation-name: test) {}")).toBe("@supports (animation-name: test) ")
+  expect(postcssCheck(
```
Comment (davidtheclark) [inline src/utils/__tests__/beforeBlockString.test.js]: I don't see an assertion here ... oh, now I see it. Could you put longer strings like this in variables so the `expect(..).toBe(..)` is more apparent?

File: src/utils/__tests__/blurInterpolation.test.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+it("blurInterpolation", () => {
+  expect(blurInterpolation("#{$selector}")).toBe(" $selector ")
+  expect(blurInterpolation("#{$selector}", "`")).toBe("`$selector`")
+  expect(blurInterpolation("#{$selector * 10px}")).toBe(" $selector * 10px ")
+  expect(blurInterpolation("#{$font-size}/#{$line-height}")).toBe(" $font-size / $line-height ")
+  expect(blurInterpolation("url(#{$selector * 10px})")).toBe("url( $selector * 10px )")
+  expect(blurInterpolation("calc(#{$selector} * 2)")).toBe("calc( $selector  * 2)")
+  expect(blurInterpolation(
+    "filter: progid:DXImageTransform.Microsoft.gradient(enabled='false', startColorstr='#{ie-hex-str($green)}', endColorstr='#{ie-hex-str($translucent-red)}');")).toBe(
```
Comment (davidtheclark) [inline src/utils/__tests__/blurInterpolation.test.js]: Same thing about the visibility of `expect(..).toBe(..)`.

File: src/utils/__tests__/declarationValueIndex.test.js [hunk 4]
```diff
@@ -1,24 +1,21 @@
...
   })
   rules("a { a  : b }", decl => {
-    t.equal(declarationValueIndex(decl), 5)
+    expect(declarationValueIndex(decl)).toBe(5)
   })
   rules("a { a:\nb }", decl => {
-    t.equal(declarationValueIndex(decl), 3)
+    expect(declarationValueIndex(decl)).toBe(3)
   })
 })
```
Comment (davidtheclark) [inline src/utils/__tests__/declarationValueIndex.test.js]: These tests have the same async issue as mentioned above.

File: src/utils/__tests__/isKeyframeRule.test.js [hunk 5]
```diff
@@ -1,40 +1,37 @@
...
   })
   rules("@media print { a {} }", rule => {
-    t.notOk(isKeyframeRule(rule), "@media")
+    expect(isKeyframeRule(rule)).toBeFalsy()
   })
   rules("@supports (animation-name: test) { a {} }", rule => {
-    t.notOk(isKeyframeRule(rule), "@supports")
+    expect(isKeyframeRule(rule)).toBeFalsy()
   })
 })
```
Comment (davidtheclark) [inline src/utils/__tests__/isKeyframeRule.test.js]: Same async problem.

File: src/utils/__tests__/isStandardSyntaxAtRule.test.js [hunk 6]
```diff
@@ -2,58 +2,55 @@ import isStandardSyntaxAtRule from "../isStandardSyntaxAtRule"
...
 import postcss from "postcss"
 import scss from "postcss-scss"
-import test from "tape"
-
-test("isStandardSyntaxAtRule", t => {
-  t.plan(12)
 
+it("isStandardSyntaxAtRule", () => {
   atRules("@charset UTF-8;", atRule => {
-    t.ok(isStandardSyntaxAtRule(atRule), "non nested at-rules without quotes")
```
Comment (davidtheclark) [inline src/utils/__tests__/isStandardSyntaxAtRule.test.js]: In all cases where we *have* descriptions for assertions, we should preserve them. This means changing the `it()` above to `describe()`, and then wrapping these assertions in `it("description", () => {..})`.

File: src/utils/__tests__/isStandardSyntaxDeclaration.test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  lessRules("@var: b", decl => {
+    expect(isStandardSyntaxDeclaration(decl)).toBeFalsy()
+  })
+  lessRules("a { @var: b }", decl => {
+    expect(isStandardSyntaxDeclaration(decl)).toBeFalsy()
+  })
+})
+
+function rules(css, cb) {
+  postcss().process(css).then(result => {
```
Comment (davidtheclark) [inline src/utils/__tests__/isStandardSyntaxDeclaration.test.js]: Async problem.

File: src/utils/__tests__/isStandardSyntaxFunction.test.js [hunk 8]
```diff
@@ -1,25 +1,22 @@
...
+    expect(isStandardSyntaxFunction(func)).toBeTruthy()
   })
 
   rules("a { $list: (list) }", func => {
-    t.notOk(isStandardSyntaxFunction(func), "scss list")
+    expect(isStandardSyntaxFunction(func)).toBeFalsy()
   })
 
   rules("a { $map: (key: value) }", func => {
-    t.notOk(isStandardSyntaxFunction(func), "scss map")
```
Comment (davidtheclark) [inline src/utils/__tests__/isStandardSyntaxFunction.test.js]: Losing descriptions.

**Comment by m-allanson** [comment]
@davidtheclark Very good points, thanks for catching those. I've added [one more commit](https://github.com/stylelint/stylelint/pull/2085/commits/968a559f19d2c7ce9e88e650262ea4971f0853b9) with an example of handling the async tests properly. Does that look like a viable pattern to use?

If you're running the code locally you can uncomment the `delayPromise` code to verify that tests are only passing after the promises have resolved.

**Comment by davidtheclark** [comment]
@m-allanson That looks great to me!

**Comment by m-allanson** [comment]
@davidtheclark I'm gradually amending these files. Once finished I'll close this PR and open a fresh one into `master`. In the meantime, let me know if you spot any issues with my recent commits here.

**Comment by davidtheclark** [comment]
@m-allanson Would you mind pointing me at specific lines you might have questions about? That's a lot of commits, and there are a lot of tests.

**Comment by m-allanson** [comment]
@davidtheclark Having just tried to pick out some parts now, I think this will be easier once I've finished and opened a new PR. I should have something up later this week.

**Pull request closed without merging**